### ISSUE TITLE
Ignore commonly generated folders

### ISSUE THREAD

**Issue by astorije** [prettier/prettier#4547]
At the moment, Prettier ignores `node_modules`. In my experience, folders `coverage` and `dist` are commonly used to contain generated artifacts and could be ignored accordingly, as a sensible default.

If someone _really_ needs to include those, adding `!coverage` and/or `!dist` in `.prettierignore` should re-include them.

Thoughts? :)

**Comment by BuckyMaler** [comment]
I think all entries in `.gitignore` should be ignored by Prettier without the need to specify `--ignore-path .gitignore`. The `.gitignore` should be a constant with the option to additionally use a `.prettierignore`. I think this is sensible because I cannot think of a reason as to why a file that is not version controlled would need to be formatted. With this, many projects may not even need to add a `.prettierignore`. A `.prettierignore` is particularly useful for ignoring files that are version controlled because that cannot be done with a `.gitignore`.

**Comment by danihodovic** [comment]
Any updates on this?

**Comment by evilebottnawi** [comment]
We can implement:
```
--ignore-pattern [String]      Pattern of files to ignore (in addition to those in .prettierignore)
```

eslint has this flag too

**Comment by danihodovic** [comment]
Is it possible to specify this in .prettierrc.yml instead of passing it as a CLI option every time?

**Comment by lydell** [comment]
@danihodovic Sounds like you’d be interested in https://github.com/prettier/prettier/issues/7073

**Comment by lydell** [comment]
Similar issue: #4081

**Comment by jdufresne** [comment]
> I think all entries in .gitignore should be ignored by Prettier without the need to specify --ignore-path .gitignore. 

Came here to make the same suggestion. I think that would reduce the need to configure prettier in the common case.

Other opinionated tools, such as [black](https://github.com/psf/black/commit/e9d4e7b67f359114ba57ecf0c1300f7eb4f08950), started doing this and it has worked great.

**Comment by alexilyaev** [comment]
I think it's safe to say that eventually Prettier will face the same requests from the community as ESLint did.
So I suggest following ESLint's ignore features:
https://eslint.org/docs/user-guide/configuring#ignoring-files-and-directories

### Specifically:
- Supporting [--ignore-pattern](https://eslint.org/docs/user-guide/command-line-interface#ignore-pattern) CLI command (can be used multiple times) to add custom ignore patterns on top of the ones from `--ignore-path`.
  - This is useful for running Prettier without needing to add config files.
  - e.g. `npx prettier --ignore-pattern ...` would be really nice.
  - But this is not comfortable when you have many patterns you want to ignore, and would require extra integration from IDE's (otherwise "Format On Save" could format a file we want to ignore).
  - Another interesting use case: Un-ignore a pattern, thus force formatting files that are usually ignored (e.g. "Format this file with Prettier" command in IDE's doesn't work without it if the file is ignored elsewhere).
- Supporting `ignorePatterns` property in Prettier configurations.
  - In ESLint, this one took several years of discussions and tickets from the community until they were finally convinced to add it.
  - This essentially solves most custom ignore use cases, as Prettier already uses the config no matter how you run it. So IDE support is out-of-the-box.
- Worth mentioning they also have [--no-ignore](https://eslint.org/docs/user-guide/command-line-interface#no-ignore)

An alternative would be to always include `.prettierignore` and let `--ignore-path` merge on top of that. Or support several `--ignore-path` commands.
But that would be confusing IMO.

**Comment by google-api-typings-generator** [comment]
I would also like to have prettier use .gitignore with ability to extend it specifically for prettier.
For example, I have a `composer.lock` file (I know, it's PHP and I'm getting rid of it, shame on me).
Prettier recognises that it's a JSON and formats it. But I don't want to, I want to ignore it.
Also, I don't want to maintain two very similar ignore files.
Ideally, there should be a way to include .gitignore into .prettierignore and extend it. Cheers!

**Comment by Maxim-Mazurok** [comment]
I would also like to have prettier use .gitignore with ability to extend it specifically for prettier.
For example, I have a composer.lock file (I know, it's PHP and I'm getting rid of it, shame on me).
Prettier recognises that it's a JSON and formats it. But I don't want to, I want to ignore it.
Also, I don't want to maintain two very similar ignore files.
Ideally, there should be a way to include .gitignore into .prettierignore and extend it. Cheers!

**Comment by jasonkarns** [comment]
I strongly agree that some form of opinionated ignores would be valuable. no-config tools are the best.

However, I urge caution if prettier were to go down the route of using .gitignore. silver-searcher has a host of bugs related to their decision to respect .gitignore; and npm itself has additional complexity in its ignore logic.

Some things to think about if prettier respected .gitignore:

- would it also respect the user-level gitignore? Likely not for the same reason that prettier doesn't support global configs. However, since many platform-level ignores are defined in system-gitignore, now there is divergent behavior from git itself. (And presumably the desire to use .gitignore is to: "only format files versioned in git". That behavior would then be violated.)

- similarly, would sub-folder .gitignores be respected?

- would prettier parse the .gitignore itself? If so, fully compliance with git's parsing rules would be necessary. (ie, negations, rooted vs relative paths, etc).

- an alternative implementation would be to essentially shell-out to git to find the list of versioned files. This avoids incompatibilities with git's parsing, but comes with its own downsides.

**Comment by Laruxo** [comment]
I would really like to see `ignorePatterns` config option. When CLI, editor and pre-commit hook runs Prettier, it requires a lot of configuration set `ignore-path` CLI option for all of these. It also makes it harder to create a shareable Prettier config.

**Comment by backbone87** [comment]
> an alternative implementation would be to essentially shell-out to git to find the list of versioned files.

i think this is a good approach and can be gated by an option `vcs`, which, if set, tries to detect the vcs used in the current WD (or errors if there is none) and then calls out to the VCS' CLI to find a list of (to be) versioned files. using this option would not change how .prettierignore and all other include/exclude options would work. it basically just replaces the file/directory listing obtained from using the FS calls to the listings that the VCS gives.

the question is how IDE intergrations would treat this. my opinion on this is, that i would expect to not see formatting errors reported for unversioned files, but still be able to apply prettier formatting explicitly.

bloating project specific `.*ignore` files by ignoring everything what different local dev setups may produce is cumbersome & errorprone. also prettier as well as eslint say in their docs that their specific ignore files behave like gitignore files. this is true, but only applies to the syntax and interpretation of the contents of a single ignore file and not --how a user would expect-- the whole ignore algorithm how it is done by git (including project local ignore, user ignore & nested ignores).

**Comment by Maxim-Mazurok** [comment]
I think that the general approach (supporting different VCSs and not just Git) is good.
It may present its challenges, however. Like for TFVC, you may have to specify the path to VCS executable, etc. (see how it's configured in VS Code plugin for TFS).
Also, sometimes devs use TFVC + Git. In my experience, because the company uses TFS but you don't like it and want to use Git. So, there might be two VCSs in the same directory.
So, I think that the best approach would be to add `--vcs` option with a default value of `--vcs=git`. And only add git support for the first iteration. And other VCSs can build on that, maybe also providing new options, such as `--tfs-exe-path="..."`, etc.

**Comment by fzf404** [comment]
When using a personal prettier configuration I found a solution.

[fzf404/prettier-config](https://github.com/fzf404/prettier-config)

```jsonc
overrides: [
	{
		files: 'pnpm-lock.yaml',
		options: {
			rangeEnd: 0, // default: Infinity
		},
	},
]
```

Use this command to ignore other file.

```jsonc
"scripts": {
	// ...
	"format": "prettier --ignore-path .gitignore --write ."
},
```

**Comment by nozomiishii** [comment]
@fzf404 
Genius! I was in trouble and you saved the day! Thanks!

**Comment by Maxim-Mazurok** [comment]
> I would also like to have prettier use .gitignore with ability to extend it specifically for prettier. For example, I have a composer.lock file (I know, it's PHP and I'm getting rid of it, shame on me). Prettier recognises that it's a JSON and formats it. But I don't want to, I want to ignore it. Also, I don't want to maintain two very similar ignore files. Ideally, there should be a way to include .gitignore into .prettierignore and extend it. Cheers!

Hey guys, I implemented extension/overrides to gitignore using unit tests.
The basic idea is to have `.gitignore` contents copied at the beginning of `.prettierignore` (and `.eslintignore`), and allow to add something else to the bottom of the file. Keeping in sync is guaranteed by tests. Perhaps updates could be automated as well.

`.gitignore`:
```ignore
node_modules
temp
etc
src/api
```

`.prettierignore`:
```ignore
### Below is the exact copy of .gitignore file, keep them in sync (tested by ignore-files.spec.ts). Having this file is required for proper IDE integration

node_modules
temp
etc
src/api

### Above is the exact copy of .gitignore file
### Below are overrides

!/src/api
```

`ignore-files.spec.ts`:
```ts
import { readFile } from "fs/promises";

[".eslintignore", ".prettierignore"].map((ignoreFileName) =>
  it(`${ignoreFileName} contains .gitignore`, async () => {
    // all the splitting/joining lines stuff is needed to make it EOL-agnostic
    const ignoreFileContent = (await readFile(ignoreFileName, "utf8"))
      .split(/[\r\n]+/)
      .filter((x) => x.trim() !== "")
      .join("\n");
    const gitIgnoreLines = (await readFile(".gitignore", "utf8")).split(/[\r\n]+/).filter((x) => x.trim() !== "");
    const expectedContent = [
      "### Below is the exact copy of .gitignore file, keep them in sync (tested by ignore-files.spec.ts). Having this file is required for proper IDE integration",
      ...gitIgnoreLines,
      "### Above is the exact copy of .gitignore file",
      "### Below are overrides",
    ].join("\n");
    expect(ignoreFileContent).toContain(expectedContent);
  })
);
```

This issue is no longer relevant to me, so I'm unsubscribing, cheers!

**Comment by U-4-E-A** [comment]
Is there a CLI flag to _ignore_ the contents of `.gitignore`? I want to do a one time formatting of my dist folder after builds. Thanks.

**Comment by Mister-Hope** [comment]
Checkout https://github.com/prettier/prettier/issues/17174#issuecomment-2684966631. My opinion is that `pnpm` and `yarn` should be contain as Node.js native things due to the introduction of CorePack.

Now every project can be forced using `pnpm` or `yarn` with `packageManager` field at root `package.json`, so if prettier can not recognize these pm's lockfiles and assets, then it shall be a bug to prettier.

Specially mention @fisker  and @kachkaev (who comoment https://github.com/prettier/prettier/issues/14517#issuecomment-1475198436 which compares pnpm and frameworks with no sense)

**Comment by sdavids** [comment]
Commonly used lock files should be excluded by default; they are usually not in `.gitignore`:

`package-lock.json` 
`pnpm-lock.yaml`
`yarn.lock`
`.devcontainer/devcontainer-lock.json`

**Comment by HatemMn** [comment]
> Commonly used lock files should be excluded by default; they are usually not in `.gitignore`:
> 
> `package-lock.json` `pnpm-lock.yaml` `yarn.lock` `.devcontainer/devcontainer-lock.json`

I agree with this comment, if someone can please tell me: is there a valid use case where formatting a lock file is relevant ? Shall is be npm, pnpm or yarn, all of them are typically out of prettier range and most devs and repos I saw create a .prettierignore file just for this
